### PR TITLE
Line symbols: Review combination of dashed line with pointed line caps

### PR DESCRIPTION
**Pull request by dg0yt**
In analysis of #1125, I stumbled over the implementation of pointed line caps, as used by the "Erosion gully" symbol. Generating the actual graphic primitives for dashed line features is one of the more complex parts of code in Mapper, taking quite some CPU and memory resources.

Given that:
- ISOM et al. don't need the combination of dashed line with pointed line caps,
- in Mapper, *pointed line caps already don't work properly with dashed lines* (cf. screenshot),
- in Mapper, pointed line caps behave different from other line caps (i.e. pointed line caps don't extend to the outside of dash lines, resulting in a dash layout which differs from common line caps),
- the layout is even more complex in the presence of "dash points",
- in graphics frameworks such as Qt, PDF and SVG, pointed line caps do not exist,
- at least some variants of dashed lines with pointed line caps might be implemented with "mid symbols" instead (similar to dot patterns),
- export to OCD format will not support the combination of dashed line styles with pointed line caps,

I propose to restrict the support of pointed line ends to continuous lines only, and as an alternative to the drawn (visible) line having an (maybe zero) offset from the start and end of the line feature (#1125). This change would:
- ensure the consistency of the layout of dashed line features,
- simplify the implementation,
- facilitate the export to OCD format.

Are there significant objections?

(This PR doesn't yet include the proposed change.)

cap-variants

### CODE REVIEW

**Comment by ollesmaps** [comment]
I like the idea of zero offset from the start and end. However with hill/ depression contours it does not work well: 
hill_contour

**Comment by dg0yt** [comment]
Zero offset will leave closed paths closed, so no problem with closed contour lines or even closed cliffs.

Non-zero positive offsets and pointed line ends are quite pointless for closed paths IMO, but IIRC the OCAD Viewer does handle this differently. Maybe you can test in OCAD a closed Gully or Cliff (#1125) and see if what you get is what you want .

**Comment by Symbian9** [comment]
> ISOM et al. don't need the combination of dashed line with pointed line caps,

We don't know what could be added in next revisions of ISOM, so if something noe work, lets (at least) just keep it dissabled.

**Comment by dg0yt** [comment]
This is where the changes lead so far:
 

As I wrote in the last commit, pure offsets at path start and end (#1125) can be implemented as a minimal variation, by simply not drawing the pointed ends but otherwise using the same layout. Of course with different length at start and end.

File: src/core/symbols/line_symbol.cpp [hunk 1]
```diff
@@ -967,20 +977,18 @@ void LineSymbol::processDashedLine(
...
-	
-	auto groups_start = SplitPathCoord::begin(path_coords);
+	auto groups_start = start;
 	auto line_start   = groups_start;
 	for (bool is_part_end = false; !is_part_end; )
 	{
 		auto groups_end_path_coord_index = path_coords.findNextDashPoint(groups_start.path_coord_index);
 		auto groups_end_index = path_coords[groups_end_path_coord_index].index;
-		auto groups_end = SplitPathCoord::at(path_coords, groups_end_path_coord_index);
+		auto groups_end = groups_end_index < end.index ? SplitPathCoord::at(path_coords, groups_end_path_coord_index) : end;
```
Comment (dg0yt) [inline src/core/symbols/line_symbol.cpp]: This turned out to be to simple. Revised version of this commit is coming.

File: src/core/symbols/line_symbol.cpp [hunk 2]
```diff
@@ -1298,11 +1308,11 @@ void LineSymbol::createMidSymbolRenderables(
...
 	}
 	
-	auto part_end = path.last_index;
-	while (groups_start.index != part_end)
+	while (groups_start.index != end.index)
 	{
 		auto groups_end_path_coord_index = path_coords.findNextDashPoint(groups_start.path_coord_index);
-		auto groups_end = SplitPathCoord::at(path_coords, groups_end_path_coord_index);
+		auto groups_end_index = path_coords[groups_end_path_coord_index].index;
+		auto groups_end = groups_end_index < end.index ? SplitPathCoord::at(path_coords, groups_end_path_coord_index) : end;
```
Comment (dg0yt) [inline src/core/symbols/line_symbol.cpp]: This turned out to be to simple. Revised version of this commit is coming.

File: src/core/symbols/symbol.cpp [hunk 3]
```diff
@@ -453,7 +453,7 @@ QImage Symbol::createIcon(const Map& map, int side_length, bool antialiasing, qr
 			}
 			else if (line->getCapStyle() == LineSymbol::PointedCap)
 			{
-				line_length_half = std::max(line_length_half, 0.0012 * line->getPointedCapLength());
+				line_length_half = std::max(line_length_half, 0.006 * (line->startOffset() + line->endOffset()));
```
Comment (dg0yt) [inline src/core/symbols/symbol.cpp]: 0.0012 / 2 = 0.0006. Fix will be force-pushed.

**Comment by Symbian9** [comment]
> LineSymbol: Add support for offset at start/end

So now Mapper has option to define different offset for line end & start, right?

If so, it would be useful for my request ;-)

- https://github.com/OpenOrienteering/mapper/issues/1111#issuecomment-403269920

**Comment by dg0yt** [comment]
> So now Mapper has option to define different offset for line end & start, right?

Once this is merged, yes.

**Comment by lpechacek** [comment]
> dg0yt requested a review from lpechacek 2 days ago

Will have to get up to speed on this one as I didn't follow the conversation closely. Will try to finish the icon pull request first and then I'll get back on this one.

**Pull request merged**